horovod/common/operations.cc

```diff
@@ -1487,11 +1337,251 @@ void BackgroundThreadLoop(HorovodGlobalState& state) {
+    std::this_thread::sleep_for(sleep_duration);
+  }
+  state.last_cycle_start = std::chrono::steady_clock::now();
+
+  // Copy the data structures from global state under this lock.
+  // However, don't keep the lock for the rest of the loop, so that
+  // enqueued stream callbacks can continue.
+  std::queue<MPIRequest> message_queue;
+  {
+    std::lock_guard<std::mutex> guard(state.mutex);
+    while (!state.message_queue.empty()) {
+      MPIRequest message = state.message_queue.front();
+      state.message_queue.pop();
+      message_queue.push(message);
+    }
+  }
+
+  // Collect all tensors that are ready to be reduced. Record them in the
+  // tensor count table (rank zero) or send them to rank zero to be
+  // recorded (everyone else).
+  std::vector<std::string> ready_to_reduce;
+  if (is_coordinator) {
+    while (!message_queue.empty()) {
+      // Pop the first available message message
+      MPIRequest message = message_queue.front();
+      message_queue.pop();
+
+      bool reduce =
+          IncrementTensorCount(state.message_table, message, state.size);
+      if (reduce) {
+        ready_to_reduce.push_back(message.tensor_name());
+      }
+    }
+
+    // Rank zero has put all its own tensors in the tensor count table.
+    // Now, it should count all the tensors that are coming from other
+    // ranks at this tick.
+
+    // 1. Get message lengths from every rank.
+    auto recvcounts = new int[state.size];
+    recvcounts[0] = 0;
+    MPI_Gather(MPI_IN_PLACE, 1, MPI_INT, recvcounts, 1, MPI_INT, RANK_ZERO,
+               state.mpi_comm);
+
+    // 2. Compute displacements.
+    auto displcmnts = new int[state.size];
+    size_t total_size = 0;
+    for (int i = 0; i < state.size; i++) {
+      if (i == 0) {
+        displcmnts[i] = 0;
+      } else {
+        displcmnts[i] = recvcounts[i - 1] + displcmnts[i - 1];
+      }
+      total_size += recvcounts[i];
+    }
+
+    // 3. Collect messages from every rank.
+    auto buffer = new char[total_size];
+    MPI_Gatherv(nullptr, 0, MPI_BYTE, buffer, recvcounts, displcmnts, MPI_BYTE,
+                RANK_ZERO, state.mpi_comm);
+
+    // 4. Process messages.
+    for (int i = 1; i < state.size; i++) {
+      std::string received_data(buffer + displcmnts[i], (size_t)recvcounts[i]);
+
+      MPIRequestList received_message_list;
+      MPIRequestList::ParseFromString(received_message_list, received_data);
+      for (auto& received_message : received_message_list.requests()) {
+        auto& received_name = received_message.tensor_name();
+
+        bool reduce = IncrementTensorCount(state.message_table,
+                                           received_message, state.size);
+        if (reduce) {
+          ready_to_reduce.push_back(received_name);
+        }
+      }
+      if (received_message_list.shutdown()) {
+        // Received SHUTDOWN request from one of the workers.
+        state.shut_down = true;
```

Comment:
break?

Reply:
We still process last tensors received in this batch.